lib/api/api.go

```diff
@@ -1617,6 +1646,30 @@ func toJsonFileInfoSlice(fs []db.FileInfoTruncated) []jsonFileInfoTrunc {
 	return res
 }
 
+func toJsonPendingDeviceSlice(devices map[protocol.DeviceID]db.ObservedDevice) []jsonPendingDevice {
+	res := make([]jsonPendingDevice, 0, len(devices))
+	for id, meta := range devices {
+		res = append(res, jsonPendingDevice{id.String(), meta})
+	}
+	return res
+}
+
+func toJsonPendingFolderSlice(folders map[string]map[protocol.DeviceID]db.ObservedFolder) []jsonPendingFolder {
+	res := make([]jsonPendingFolder, 0, len(folders))
+	for id, devices := range folders {
+		res = append(res, jsonPendingFolder{id, toJsonOfferingDeviceSlice(devices)})
+	}
+	return res
+}
+
+func toJsonOfferingDeviceSlice(devices map[protocol.DeviceID]db.ObservedFolder) []jsonOfferingDevice {
+	res := make([]jsonOfferingDevice, 0, len(devices))
+	for id, meta := range devices {
+		res = append(res, jsonOfferingDevice{id.String(), meta})
+	}
+	return res
+}
```

Comment:
The API was discussed [here](https://github.com/syncthing/docs/pull/498) before getting implemented.  I can't remember a reason why it was designed as an array instead of an object with device IDs as key / attribute name.  I think it will be much simpler and mirrors existing endpoints (`GET /rest/stats/*` and `GET /rest/system/connections`), so working on this now.

Reply:
One thing that justifies the conversion here is that copying a `jsonObservedDevice` entry to a `config.ObservedDevice` entry for example is easier when the former also has the `deviceID` attribute in JSON encoding.  Same goes for folders.

So the complexity is really just shifted to the GUI code doing the restucturing.  That is probably an acceptable trade-off though?

Reply:
> `protocol.DeviceID` implements `MarshalString`

What's relevant here is rather the `MarshalText()` method.  That is in fact implemented, but with a pointer receiver.  If I change that to a simple value receiver, the method is called and works in JSON serialization.  Don't know what other consequences might be though.

Reply:
Implemented in acf9c0322a808734025c52ca9235f92493af1b74.